### PR TITLE
KAA-1203: Support failover when KAATCP_CONNACK_REFUSE_BAD_CREDENTIALS is received

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @forGGe @rasendubi

### CODE REVIEW

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 1]
```diff
@@ -41,13 +41,17 @@ typedef struct {
     uint16_t    channel_count;
 } kaa_sync_info_t;
 
-
+typedef struct {
+    kaa_auth_failure_fn callback;
+    void                *context;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]:
``` c
    kaa_auth_failure_fn callback;
    void               *context; // note alignment here
```

Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Anyway, I wouldn't align at all.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 2]
```diff
@@ -41,13 +41,17 @@ typedef struct {
...
+} kaa_auth_failure_handler_t;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Don't use `_t` as a suffix.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 3]
```diff
@@ -461,3 +467,19 @@ kaa_error_t kaa_channel_manager_on_new_access_point(kaa_channel_manager_t *self
 
     return KAA_ERR_NONE;
 }
+
+void kaa_channel_manager_set_auth_failure_handler(kaa_channel_manager_t *self
+                                                , kaa_auth_failure_fn handler, void *context)
+{
+    KAA_RETURN_IF_NIL(self, );
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Don't use `KAA_RETURN_IF_NIL`

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 4]
```diff
@@ -461,3 +467,19 @@ kaa_error_t kaa_channel_manager_on_new_access_point(kaa_channel_manager_t *self
...
+    self->auth_failure_handler.callback = handler;
+    self->auth_failure_handler.context  = context;
+}
+
+void kaa_channel_manager_process_auth_failure(kaa_channel_manager_t *self
+                                            , kaa_auth_failure_reason_t reason)
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Code style

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 5]
```diff
@@ -461,3 +467,19 @@ kaa_error_t kaa_channel_manager_on_new_access_point(kaa_channel_manager_t *self
...
+{
+    KAA_RETURN_IF_NIL2(self, self->auth_failure_handler.callback, );
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Don't use `KAA_RETURN_IF_NIL2`

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 6]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
 kaa_error_t kaa_channel_manager_remove_transport_channel(kaa_channel_manager_t *self
                                                        , uint32_t channel_id);
 
+
+
+/**
+ * @brief Specifies authorization failure reason.
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: `@brief` is not needed for one sentence.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 7]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
...
+typedef enum {
+    KAA_AUTH_STATUS_UNKNOWN, /**< Authorization failed for unknown reason. */
+    KAA_AUTH_STATUS_BAD_CREDENTIALS, /**< Authorization failed because credentials are invalid. */
+    KAA_AUTH_STATUS_VERIFICATION_FAILED /**< Authorization failed because of verification failure. */
+} kaa_auth_failure_reason_t;
+
+
+
+/**
+ * @brief Processes authorization failure.
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: `@brief` is not needed for one sentence.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 8]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
...
+
+/**
+ * @brief Specify authorization failure handler
+ *
+ * @param[in]   self            Channel manager.
+ * @param[in]   handler         Authorization failure handler.
+ * @param[in]   context         Handler context.
+ */
+void kaa_channel_manager_set_auth_failure_handler(kaa_channel_manager_t *self
+                                                , kaa_auth_failure_fn handler, void *context);
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: code style

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 9]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
...
+
+
+/**
+ * @brief Processes authorization failure if valid handler exists.
+ *
+ * @param[in]   self            Channel manager.
+ * @param[in]   reason          Authorization failure reason.
+ */
+void kaa_channel_manager_process_auth_failure(kaa_channel_manager_t *self
+                                            , kaa_auth_failure_reason_t reason);
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: code style

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 10]
```diff
@@ -96,10 +100,12 @@ kaa_error_t kaa_channel_manager_create(kaa_channel_manager_t **channel_manager_p
     (*channel_manager_p)->transport_channels      = kaa_list_create();
     KAA_RETURN_IF_NIL((*channel_manager_p)->transport_channels, KAA_ERR_NOMEM);
 
-    (*channel_manager_p)->kaa_context             = context;
-    (*channel_manager_p)->sync_info.request_id    = 0;
-    (*channel_manager_p)->sync_info.is_up_to_date = false;
-    (*channel_manager_p)->logger                  = context->logger;
+    (*channel_manager_p)->kaa_context                   = context;
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Avoid using ugly constructs, like `(*channel_manager_p)`

Just declare above pointer-to-object and use it

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 11]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
...
+
+
+/**
+ * @brief Specifies authorization failure reason.
+ *
+ */
+typedef enum {
+    KAA_AUTH_STATUS_UNKNOWN, /**< Authorization failed for unknown reason. */
+    KAA_AUTH_STATUS_BAD_CREDENTIALS, /**< Authorization failed because credentials are invalid. */
+    KAA_AUTH_STATUS_VERIFICATION_FAILED /**< Authorization failed because of verification failure. */
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: Add trailing comma.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 12]
```diff
@@ -96,10 +100,12 @@ kaa_error_t kaa_channel_manager_create(kaa_channel_manager_t **channel_manager_p
...
+    (*channel_manager_p)->sync_info.request_id          = 0;
+    (*channel_manager_p)->sync_info.is_up_to_date       = false;
+    (*channel_manager_p)->logger                        = context->logger;
+    (*channel_manager_p)->auth_failure_handler.callback = NULL;
+    (*channel_manager_p)->auth_failure_handler.context  = context;
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Copy-paste detected: you are assigning wrong context. `NULL` here is more appropriate

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h [hunk 13]
```diff
@@ -101,6 +101,50 @@ kaa_transport_channel_interface_t *kaa_channel_manager_get_transport_channel(kaa
...
+} kaa_auth_failure_reason_t;
```
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: Don't use `_t` as a suffix.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c [hunk 14]
```diff
@@ -461,3 +467,19 @@ kaa_error_t kaa_channel_manager_on_new_access_point(kaa_channel_manager_t *self
...
+
+void kaa_channel_manager_set_auth_failure_handler(kaa_channel_manager_t *self
+                                                , kaa_auth_failure_fn handler, void *context)
+{
+    KAA_RETURN_IF_NIL(self, );
+    self->auth_failure_handler.callback = handler;
+    self->auth_failure_handler.context  = context;
+}
+
+void kaa_channel_manager_process_auth_failure(kaa_channel_manager_t *self
```
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: You are violating style guide rules. Do not put comma on next line

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c (hunk 4 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Indent violates style-guide. Use source formatting script

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c (hunk 12 above)
Comment (rasendubi) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Use an intermediate variable to not dereference `channel_manager_p` all the time.

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.c (hunk 5 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.c]: Again, `*RETURN_IF_*` is prohibited

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h (hunk 8 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: Style

File: client/client-multi/client-c/src/kaa/kaa_channel_manager.h (hunk 9 above)
Comment (forGGe) [inline client/client-multi/client-c/src/kaa/kaa_channel_manager.h]: Style

**Comment by RostakaGmfun** [comment]
@forGGe re-review, please

**Comment by forGGe** [comment]
+

**Comment by RostakaGmfun** [comment]
@rasendubi ?

**Comment by rasendubi** [comment]
LGTM

**Pull request merged**